jax/_src/nn/functions.py

```diff
@@ -765,3 +767,174 @@ def hard_silu(x: ArrayLike) -> Array:
+  # Mask
+  if mask is not None:
+    if mask.dtype == jnp.bool_:
+      large_negative_number = _get_large_negative(logits.dtype)
+      padded_logits = jnp.where(mask, logits, large_negative_number)
+    else:
+      padded_logits = logits + mask
+  else:
+    padded_logits = logits
+
+  if is_causal:
+    T, S = query.shape[-3], key.shape[-3]
+    assert T == S, f'T should equal to S, but got {T} vs {S}'
+    mask = _get_causal_mask(T, logits.dtype)
+    padded_logits = padded_logits + mask
+
+  # Softmax and it is always carried out in fp32.
+  padded_logits = padded_logits.astype(jnp.float32)
+  probs = jax.nn.softmax(padded_logits, axis=-1).astype(key.dtype)
+
+  # Compute the attention context.
+  encoded = jnp.einsum('BNTS,BSNH->BTNH', probs, value)
+  return encoded, probs
+
+def dot_product_attention(
+    query: ArrayLike,
+    key: ArrayLike,
+    value: ArrayLike,
+    bias: ArrayLike | None = None,
+    mask: ArrayLike | None = None,
+    *,
+    scale: float | None = None,
+    is_causal: bool = False,
+    implementation: str | None = None) -> Array:
+  r"""Scaled dot product attention function.
+
+  Computes the attention function on Query, Key, and Value tensors:
+
+  .. math ::
+    \mathrm{Attention}(Q, K, V)=\mathrm{softmax}(\frac{QK^T}{\sqrt{d_k}}V)
+
+  If we define :code:`logits` as the output of :math:`QK^T` and the
+  :code:`probs` as the output of :math:`softmax`.
+
+  Throughout this function, we utilize the following uppercase letters to
+  represent the shape of array:
+
+    B = batch size
+    S = length of the key/value (source)
+    T = length of the query (target)
+    N = number of attention heads
+    H = dimensions of each attention head
+
+  Args:
+    query: query array; shape :code:`(BTNH)`
+    key: key array; shape :code:`(BSNH)`
+    value: value array; shape :code:`(BSNH)`
+    bias: optional, bias array to be added to logits; shape broadcastable to
+          :code:`(BNTS)`.
+    mask: optional, mask array used to filter out logits. Two types of masks are
+          supported. (1) A boolean mask where `True` indicates the element
+          should take part in attention; (2) A float mask of the same type as
+          query that will be added to the attention score, meaning 0. represents
+          `True` and a large negative number 
+          (e.g. :code:`-0.5 * jnp.finfo(dtype).max`) represents `False`. The
+          shape is broadcastable to :code:`(BNTS)`.
+    scale: scale for the logits. If None, the scale will be set to 1 divided by
+           the square root of query size. Note, the currently implementation
+           only supports post-bias scale.
+    is_causal: If true, assumes upper left causal attention masking.
+    implementation: A string to control which implementation to use. Supported
+                    strings are `xla` (default), `cudnn` (cuDNN flash
+                    attention).
+
+  Returns:
+    An array with the same shape of :code:`query`. If flash attention is not
```

Comment:
The output type signature should not depend on the `implementation`. Why do we need to return `probs`? If this becomes necessary, should probably control this with an argument and fail if `implementation='cudnn'`.

Reply:
Sure. The new change makes the API only returns the attention output. Added a new `return_probs` parameter and it will return `probs` when it is True.